src/libraries/Common/src/System/Net/SocketAddress.cs

```diff
@@ -179,51 +202,29 @@ internal IPEndPoint GetIPEndPoint()
         internal void CopyAddressSizeIntoBuffer()
         {
             int addressSizeOffset = GetAddressSizeOffset();
-            Buffer[addressSizeOffset] = unchecked((byte)(InternalSize));
-            Buffer[addressSizeOffset + 1] = unchecked((byte)(InternalSize >> 8));
-            Buffer[addressSizeOffset + 2] = unchecked((byte)(InternalSize >> 16));
-            Buffer[addressSizeOffset + 3] = unchecked((byte)(InternalSize >> 24));
+            InternalBuffer[addressSizeOffset] = unchecked((byte)(InternalSize));
+            InternalBuffer[addressSizeOffset + 1] = unchecked((byte)(InternalSize >> 8));
+            InternalBuffer[addressSizeOffset + 2] = unchecked((byte)(InternalSize >> 16));
+            InternalBuffer[addressSizeOffset + 3] = unchecked((byte)(InternalSize >> 24));
         }
 
         // Can be called after the above method did work.
         internal int GetAddressSizeOffset()
         {
-            return Buffer.Length - IntPtr.Size;
+            return InternalBuffer.Length - IntPtr.Size;
         }
 #endif
 
         public override bool Equals(object? comparand) =>
-            comparand is SocketAddress other &&
-            Buffer.AsSpan(0, Size).SequenceEqual(other.Buffer.AsSpan(0, other.Size));
+            comparand is SocketAddress other && Equals(other);
+
+        public bool Equals(SocketAddress? comparand) => comparand != null && Buffer.Span.SequenceEqual(comparand.Buffer.Span);
 
         public override int GetHashCode()
         {
-            if (_changed)
-            {
-                _changed = false;
-                _hash = 0;
-
-                int i;
-                int size = Size & ~3;
-
-                for (i = 0; i < size; i += 4)
-                {
-                    _hash ^= BinaryPrimitives.ReadInt32LittleEndian(Buffer.AsSpan(i));
-                }
-                if ((Size & 3) != 0)
-                {
-                    int remnant = 0;
-                    int shift = 0;
-
-                    for (; i < Size; ++i)
-                    {
-                        remnant |= ((int)Buffer[i]) << shift;
-                        shift += 8;
-                    }
-                    _hash ^= remnant;
-                }
-            }
-            return _hash;
+            HashCode hash = default;
+            hash.AddBytes(Buffer.Span);
```

Comment:
It'd be a bit more efficient to do:
```suggestion
            hash.AddBytes(new ReadOnlySpan<byte>(InternalBuffer, 0, InternalSize));
```

Reply:
updated but I don't understand why. Is it because getting Span from away is after than from Memory or is it because of ReadOnlySpan vs Span? .... Or something else?

Reply:
Using `hash.AddBytes(Buffer.Span)` effectively expands to `hash.AddBytes((ReadOnlySpan<byte>)new Memory<byte>(InternalBuffer, 0, InternalSize).Span)`.  So we're creating a memory from an array, only to then ask for its Span.  Calling [`Memory<T>.Span`](https://github.com/dotnet/runtime/blob/89030ce9a75c98f350a13ddbe42f39a37aea1371/src/libraries/System.Private.CoreLib/src/System/Memory.cs#L266-L351) is relatively more expensive compared to just creating the span directly: `hash.AddBytes(new ReadOnlySpan<byte>(InternalBuffer, 0, InternalSize))`. We're talking nanoseconds, so it's not a huge deal, but here it's trivial to avoid, and GetHashCode can be on fast paths.

Reply:
That make sense. I forget that the Memory is created every time it is access.